### PR TITLE
L-GATr 2.0

### LINKED ISSUE #19: L-GATr 2.0 Wishlist
Add attention weight dropout support to all backends, not just sdpa [EDIT: unified attention argument handeling regardless of backend asymmetry is the dream]

also how necessary is dropping low version torch support? lots of  university HPC clusters have torch <2.4, mine included [EDIT: I do have a higher version available but it was hard to find due to .sif mislabelling, not related to below]

### PR DESCRIPTION
**Pull request by spinjo**
### Added

- `norm_elementwise_affine=True` option for all networks (changed default behavior)
- `sparse_gp=True` and `sparse_linear=True` options in `PrimitivesConfig` (changed default to `sparse_gp=True` because always faster)
- `nonlinearity_v` option for `LGATrSlim`/`ConditionalLGATrSlim` (changed default to `nonlinearity_v="sigmoid"` because more stable)
- `naive_amp=False` option and public `naive_amp` context manager to bypass `minimum_autocast_precision` and run the forward in the surrounding autocast dtype (e.g. bf16)
- `torch.compile` support for `LGATr`/`ConditionalLGATr`; `warmup_caches` helper for primitives under `mode="reduce-overhead"`
- `activation_memory_budget` option in `torch.compile` to trade backward FLOPs for a lower activation-memory peak (helps for `LGATrSlim`)
- Unit tests for all supported torch versions `torch>=2.4`; generally extended unit tests
- `embed_bivector` and `extract_bivector` interface

### Fixed

- Consistently support `scalars=None`
- Set `requires_grad_(False)` for unused params to avoid `DDP` issues
- Unified docstrings, and format equations for sphinx readability
- Micro speed/memory optimizations for `LGATr`/`LGATrSlim` primitives
- `varlen` and `xformers` attention backends now support head dims that are not a multiple of 8 via zero-padding
- Custom access to `xformers` kernels to allow `torch.compile` `xformers` attention backend without graph breaks

### Changed

- Updated `docs/` and `README.md`
- `LGATrSlim`/`ConditionalLGATrSlim` hidden layers store vectors channel-last `(..., 4, channels)`
- Unify `get_nonlinearity()` between `LGATr`/`LGATrSlim`
- Enforce `torch>=2.4` autocast syntax, drop support for low-torch-2 versions
- Unify variable names across the code; affects public API for conditional networks
- `PrimitivesConfig` is now a model input like `MLPConfig`, no global `gatr_config` anymore
- Renamed `PrimitivesConfig` flags: `use_fully_connected_subgroup`→`subgroup`, `use_bivector`→`bivector`, `use_geometric_product`→`geometric_product`
- `minimum_autocast_precision` return outputs as tuple and downcast to low dtype automatically
- Removed scalar bias from qkv linear layers in all models because redundant
- Slim stability refinements: initialize `linear_s` bias to 0, scale GLU inner product by 1/sqrt(4)
- Replaced the separate `compile_mode`/`compile_dynamic`/`compile_fullgraph` arguments with a single `compile_kwargs` dict forwarded verbatim to `torch.compile` in `compile_model` and all nets.
- Different amp strategy: vector/multivector path stays in fp32, only scalar path uses amp
- Unified the `LGATr` config names:  `increase_hidden_channels`->`attn_ratio`,  `increase_hidden_channels`->`mlp_ratio`, `activation`->`nonlinearity`, `num_hidden_layers`->`num_layers_mlp`
- Renamed slim building blocks with a `Slim` prefix (`SlimLinear` etc)

### Removed

- `einops`/`opt_einsum`/`numpy`/`lloca` requirements, now simply `torch>=2.4`

Closes #19

### CODE REVIEW

**Review by vbpla — APPROVED** [review]
I think all the changes look great. I just have some minor comments:

-  I am not sure about the factor sqrt(4) to the inner product in lgatrslim. I don't think it's fully equivalent to the attention norm because here you don't have a flat metric. I can believe that it improves performance somehow, but I don't think 4 is necessarily the optimal factor
- I would update the docs or the README adding recommendations regarding the attention backends and the compiled mode

File: README.md [hunk 1]
```diff
@@ -59,6 +59,8 @@ Overview of features in L-GATr:
 - https://github.com/heidelberg-hepml/high-dim-unfolding: Generative jet substructure unfolding with L-GATr, uses the `ConditionalLGATr`. ([paper](arxiv.org/abs/2510.19906))
 - https://github.com/gregorkrz/jetcluster: IRC-safe jet clustering with L-GATr, starting from the https://github.com/heidelberg-hepml/lorentz-gatr repo. ([paper](https://ml4physicalsciences.github.io/2025/files/NeurIPS_ML4PS_2025_59.pdf))
 - https://github.com/heidelberg-hepml/tagger-quantization: Quantized jet taggers, including float8+ternary weight implementations of L-GATr and L-GATr-slim. ([paper](https://arxiv.org/abs/2512.17011))
+- https://github.com/gregorkrz/jetclustering: Jet clustering with IRC-safety via a penalty loss term, using a L-GATr backbone.
```
Comment (AntoinePTJ) [inline README.md]: already in the list two entries above (but the correct link seems to be `https://github.com/gregorkrz/jetclustering` and not `https://github.com/gregorkrz/jetcluster` anymore)

File: .github/workflows/tests.yaml [hunk 2]
```diff
@@ -10,10 +10,30 @@ jobs:
...
+      matrix:
+        include:
+          # Python coverage at latest torch version
+          - { python-version: '3.10', torch-version: 'latest', extra: '' }
+          - { python-version: '3.11', torch-version: 'latest', extra: '' }
+          - { python-version: '3.12', torch-version: 'latest', extra: '' }
+          - { python-version: '3.13', torch-version: 'latest', extra: '' }
+
+          # Torch coverage at common python version
+          # torch<=2.4 are probably also supported, but require care with numpy install, torch.compile etc
```
Comment (AntoinePTJ) [inline .github/workflows/tests.yaml]: it should be torch<2.4 or <=2.3

File: lgatr/layers/attention/qkv.py [hunk 3]
```diff
@@ -109,123 +116,132 @@ def forward(
...
             qk_scalars = scalars
 
         # Project to queries, keys, and values (multivector reps)
         q_mv, q_s = self.q_linear(
-            qk_inputs, qk_scalars
+            qk_multivectors, qk_scalars
         )  # (..., num_items, hidden_channels * num_heads, 16)
-        k_mv, k_s = self.k_linear(qk_inputs, qk_scalars)  # (..., num_items, hidden_channels, 16)
-        v_mv, v_s = self.v_linear(inputs, scalars)  # (..., num_items, hidden_channels, 16)
+        k_mv, k_s = self.k_linear(
```
Comment (AntoinePTJ) [inline lgatr/layers/attention/qkv.py:227]: any reason why we don't use a single equilinear like multihead but with `num_heads + 2` instead of `num_heads * 3`?

File: lgatr/layers/attention/config.py [hunk 4]
```diff
@@ -11,44 +13,43 @@ class SelfAttentionConfig:
...
+        Whether to do multi-query attention. Multi-query attention decreases memory consumption
+        and parameter count by using a single set of keys and values for all heads.
+    increase_hidden_channels
         Factor by which to increase the number of hidden channels (both multivectors and scalars).
-        Vanilla transformers use 1, we use 2 for backward compatibility.
-    head_scale: bool
-        Whether to use HeadScaleMHA following the NormFormer, see https://arxiv.org/pdf/2110.09456.
-        Before combining the heads, each head is scaled by a learnable parameter.
-
+        Vanilla transformers use 1; we use 2 for backward compatibility.
```
Comment (AntoinePTJ) [inline lgatr/layers/attention/config.py]: Not sure what this means since it is set to 1 below and not 2

File: lgatr/layers/mlp/geometric_bilinears.py [hunk 5]
```diff
@@ -47,50 +48,57 @@ def __init__(
...
+            out_s_channels=0,
         )
         self.linear_right = EquiLinear(
             in_mv_channels,
             hidden_mv_channels,
+            primitives,
             in_s_channels=in_s_channels,
-            out_s_channels=None,
+            out_s_channels=0,
             initialization="almost_unit_scalar",
```
Comment (AntoinePTJ) [inline lgatr/layers/mlp/geometric_bilinears.py:63]: Why the need for a different initialization between left and right?

File: lgatr/nets/lgatr_slim.py [hunk 6]
```diff
@@ -1,118 +1,185 @@
...
+        outputs_v
+            Normalized Lorentz vectors, same shape as ``vectors``.
+        outputs_s
+            Normalized scalar features, same shape as ``scalars``.
         """
-        v_squared_norm = (vectors[..., 0].square() - vectors[..., 1:].square().sum(dim=-1)).abs()
+        v_squared_norm = (vectors.square() * self.metric).sum(-1).abs()
         s_squared_norm = scalars.square()
         total_features = v_squared_norm.shape[-1] + s_squared_norm.shape[-1]
         mean_squared_norms = (v_squared_norm.sum(-1) + s_squared_norm.sum(-1)) / total_features
```
Comment (AntoinePTJ) [inline lgatr/nets/lgatr_slim.py:168]: this is different from the equation in the paper, there is a mistake in the paper but it's also not clear to me that this formula is necessarily the correct one, there could be a better balance between vector and scalar norm? (it also links to victor's comment on the sqrt(4) scale)

File: CHANGELOG.md [hunk 7]
```diff
@@ -5,6 +5,44 @@ All notable changes to this project will be documented in this file.
...
+
+### Changed
+
+- Unify `get_nonlinearity()` between `LGATr`/`LGATrSlim`
+- Enforce `torch>=2.4` autocast syntax, drop support for low-torch-2 versions
+- Unify variable naming across the code; affects public API for conditional networks
+- `PrimitivesConfig` is now a model input like `MLPConfig`, no global `gatr_config` anymore
+- Renamed `PrimitivesConfig` flags: `use_fully_connected_subgroup`→`subgroup`, `use_bivector`→`bivector`, `use_geometric_product`→`geometric_product`
+- `minimum_autocast_precision` return outputs as tuple and downcast to low dtype automatically
+- Removed scalar bias from qkv linear layers in all models
```
Comment (AntoinePTJ) [inline CHANGELOG.md]: what's the reason for this?

**Review by AntoinePTJ — APPROVED** [review]
Everything looks good, I just have a few questions out of curiosity mostly

File: CHANGELOG.md (hunk 7 above)
Comment (spinjo) [reply CHANGELOG.md]: Because it's the convention in the definition of self-attention. For q and k one can show that an constant extra term drops out in the attention softmax. A bias in v would actually modify the result, but I think its fair to argue that a bias somewhere else does the same, and this single bias is not required. This should make the code slightly faster, but I think that gain is negligible. So its mostly a cleanness thing.

File: lgatr/layers/mlp/geometric_bilinears.py (hunk 5 above)
Comment (spinjo) [reply lgatr/layers/mlp/geometric_bilinears.py:63]: The point of the `unit_scalar` and `almost_unit_scalar` initializations in LGATr is exactly to be used here. The formal argument behind this (and also e.g. how the width of the gaussian or uniform distribution used for the weight initialization are picked) is that the activation variance should stay constant across layers. This is essential to get stable training for deep networks.

For the LGATr geometric product specifically, the idea is: You compute sum_fancy a*b, where a and b are both activations. If a and b are unit gaussians, then a*b will not be a unit gaussian, but 1*b (unit_scalar) or (1+0.1*a)*b (almost_unit_scalar) will do a better job.

LGATrSlim currently doesn't do anything equivalent in the GatedLinearUnit. I played with this when I first built LGATrSlim, but back then I dropped tuning this because empirically I didnt see a big impact, and there were other things to do. But maybe we can add this one day, would be good.

File: lgatr/layers/attention/config.py (hunk 4 above)
Comment (spinjo) [reply lgatr/layers/attention/config.py]: This is outdated. The default was 2 originally because thats more natural with multi-query, and multi-query was the default in the original GATr paper. I updated the comment.

File: lgatr/nets/lgatr_slim.py (hunk 6 above)
Comment (spinjo) [reply lgatr/nets/lgatr_slim.py:168]: I think the vector part agrees with equation 4 in the paper. The `vectors.square() * self.metric` computes the list of entries in the inner product, the `.sum(-1)` then evaluates the inner product, and we do `.abs()` in the end to make this thing positive.

But I found a typo in equation 4 in the scalar part^^

Comment (spinjo) [reply lgatr/nets/lgatr_slim.py:168]: But maybe I'm missing something. Do you have a proposal for how to combine the vector and scalar norm differently?

File: lgatr/layers/attention/qkv.py (hunk 3 above)
Comment (spinjo) [reply lgatr/layers/attention/qkv.py:227]: This is possible only if we `additional_qk_mv_channels==0`, and same for the scalars. So we could add this shortcut with an `if`, but we would also have to keep the old option. The multi-query path is not used so much, so I think its not worth it.

File: lgatr/nets/lgatr_slim.py (hunk 6 above)
Comment (AntoinePTJ) [reply lgatr/nets/lgatr_slim.py:168]: What is done in the code is summing the two and dividing by `c_v + c_s`. What is written in the overleaf is the sum of vectors divided by `c_v` + the sum of scalars divided by `c_s`. If they have a different scale it could also make sense to weight them differently (e.g. with the sqrt(4) factor for vectors, but it works as is so probably no need to change it)

File: lgatr/layers/attention/config.py (hunk 4 above)
Comment (AntoinePTJ) [reply lgatr/layers/attention/config.py]: This comment appears several times

Comment (spinjo) [reply lgatr/layers/attention/config.py]: Thanks, good catch. I also removed it in the cross-attention config. Or is there another mention?

Comment (AntoinePTJ) [reply lgatr/layers/attention/config.py]: in the mlpconfig, `Vanilla transformers use 4; we use 2 for backward compatibility.`

**Comment by spinjo** [comment]
@vbpla @AntoinePTJ I'm done with the changes, the plan is to release early next week. The one thing I still have on my list is if there's a clean way to continue support for `torch<2.4`. Compared to the June round, the main change is the `docs/` rewrite. If you have time, feel free to have a look again, but only if you have time. Thanks for the help! :)

File: lgatr/nets/__init__.py [hunk 8]
```diff
@@ -1,4 +1,17 @@
+"""Top-level L-GATr networks: full multivector variants and slim (vector + scalar) variants."""
+
 from .conditional_lgatr import ConditionalLGATr
-from .conditional_lgatr_slim import ConditionalLGATrSlim
+from .conditional_slim import ConditionalLGATrSlim
 from .lgatr import LGATr
-from .lgatr_slim import LGATrSlim
+from .slim import LGATrSlim
+from .slim_layers import (
```
Comment (AntoinePTJ) [inline lgatr/nets/__init__.py]: we could move `slim_layers` to `lgatr.layers` instead of `lgatr.nets`?

Comment (spinjo) [reply lgatr/nets/__init__.py]: Good point
